### DISCUSSION TITLE
Fuzzy check shingles algorithm

### DISCUSSION THREAD

**Discussion by mdierksen** [rspamd/rspamd#4574] [Q&A]
Can someone please explain how the algorithm for the fuzzy check with shingles currently implemented in Rspamd is working in detail?

I read the cited paper (https://www.ambuehler.ethz.ch/CDstore/www6/Technical/Paper205/Paper205.html) and the algorithm for estimated resemblance seems to be something like this:

Adding a document to fuzzy storage:

1. Determine the set of shingle hashes for the document and create a subset of the s smallest hashes.
2. For the document hash save the list of shingle hashes in the subset.
3. For each shingle hash in the subset add the document hash to the list of documents containing that shingle.

Checking document against fuzzy storage:

1. Determine the set of shingle hashes for document and create a subset S1 of the s smallest hashes.
2. For each shingle hash check whether that hash is already known and if so increase the count for all document hashes contained in the list for that hash.
3. For all document hashes with a count >= s * r (where r is the minimum resemblance we are looking for, currently r = 0.5) lookup the corresponding subset of shingle hashes S2, count the number of shingles included in S1 and S2 as well as in the subset created by selecting the s smallest hashes from the union of S1 and S2 and then calculate the resemblance by dividing that count by s.

I took a look in the code and the algorithm currently implemented in Rspamd however seems to be:

Adding a document to fuzzy storage:

1. Determine the set of shingle hashes for the document and create a subset of the s smallest hashes.
2. For each shingle hash in the subset save the document hash as well as the ordinal number in the subset.

Checking document against fuzzy storage:

1. Determine the set of shingle hashes for document and create a subset of the s smallest hashes.
2. For each shingle hash check whether that hash with that ordinal number in the subset is already known and if so determine the corresponding document hash.
3. Determine the largest number of consecutive shingle hashes in the subset belonging to the same document hash and calculate the resemblance by dividing that count by s.

Maybe I am missing something but the current implementation does not seem to be equivalent to the algorithm described in the paper. For each shingle only the document hash for the document added last is accounted for. Furthermore I do not get why the order of the hashes is important and only consecutive matches are counted. Frankly I do not see how this is working at all.

**Comment by vstakhov** [comment]
The order was added to reduce the probability of the collisions, that is quite high if we just use 64 bit values due to birthday paradox. There is also no list of the exact hashes per shingles combination due to the Redis structures limitations (and it seems that it is the only efficient way). However, there might be some other issues with the existing algorithm that I've not aware of, so I'm going to investigate this question more precisely.
